### PR TITLE
Add plugin name accessor

### LINKED ISSUE #5: Provide Name accessor for Plugin class
**Original report ([archived issue](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-plugin/issues/5)) by Steve Peters (Bitbucket: [Steven Peters](https://bitbucket.org/%7B2ccfed09-18b8-4921-8d58-15ef01092802%7D/), GitHub: [scpeters](https://github.com/scpeters)).**

----------------------------------------
Migrated from https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-common/issues/32/provide-name-accessor-for-plugin-class (ignitionrobotics/ign-common#32)

The name of a plugin is stored within the [Info struct](https://github.com/ignitionrobotics/ign-plugin/blob/94482af3e9b2cc1cbb7c2dbeb2c7846f8283f0f6/core/include/ignition/plugin/Info.hh#L51:58), but the [Plugin class](https://github.com/ignitionrobotics/ign-plugin/blob/94482af3e9b2cc1cbb7c2dbeb2c7846f8283f0f6/core/include/ignition/plugin/Plugin.hh) doesn't yet have an accessor for the name. It should be easy to add it.

### PR DESCRIPTION
**Pull request by nlamprian**
Closes #5

Signed-off-by: Nick Lamprianidis <nlamprian@gmail.com>

### CODE REVIEW

File: core/include/ignition/plugin/Plugin.hh
Comment (mxgrey) [inline core/include/ignition/plugin/Plugin.hh:141]: This function could return by const-reference (i.e. `const std::string&`) instead of by value (i.e. `std::string`) to avoid unnecessary heap allocation and copying.

**Review by mxgrey — COMMENTED** [review]
Thanks for the contribution! I just have one remark, left as an inline comment.

File: core/include/ignition/plugin/Plugin.hh
Comment (nlamprian) [reply core/include/ignition/plugin/Plugin.hh]: Valid point :)

**Review by chapulina — COMMENTED** [review]
Thanks for the PR! Mind adding some tests to check names here:

https://github.com/ignitionrobotics/ign-plugin/blob/main/test/integration/plugin.cc

And maybe also here:

https://github.com/ignitionrobotics/ign-plugin/blob/main/test/integration/templated_plugins.cc

**Comment by nlamprian** [comment]
Good thing that you asked for the tests. I removed the const reference and added a check for the pointer.

Btw last time, I had a hard time following what the classes do. These tests are a nice overview.

File: core/include/ignition/plugin/Plugin.hh
Comment (scpeters) [reply core/include/ignition/plugin/Plugin.hh:142]: can you please make this change?

Comment (nlamprian) [reply core/include/ignition/plugin/Plugin.hh:142]: Please see the comment below. I removed the reference so I can return a value in case 'info' is not initialized.

Comment (scpeters) [reply core/include/ignition/plugin/Plugin.hh:142]: > can you please make this change?

ok, I saw that you did make the change but then reverted it in case the `info` data structure is a `nullptr`. does that work for you @mxgrey?

Comment (mxgrey) [reply core/include/ignition/plugin/Plugin.hh:142]: IIRC, if `info` is ever a `nullptr` during a time that a user could call this function, then the `ignition-plugin` library is fundamentally broken somewhere.

But also I don't see where/if that requirement explicitly enforced, so perhaps it's better to not risk returning a reference.

Instead, we could return a `const std::string*` which has a `nullptr` value if `info` is a `nullptr`, or otherwise returns a pointer to the name of the plugin. This would be an equivalent way of avoiding a copy while accommodating the possibility that no `std::string` is actually available.

Comment (nlamprian) [reply core/include/ignition/plugin/Plugin.hh:142]: Ok then, I'll use a pointer. (fyi I changed it as a result of the PluginPtr.Empty test crashing)

Comment (mxgrey) [reply core/include/ignition/plugin/Plugin.hh]: Ah, okay, then I was wrong about `info` always being non-null. I'm glad that test exists :+1:

Comment (chapulina) [reply core/include/ignition/plugin/Plugin.hh]: Is the copy really a big concern here? Shouldn't the compiler take care of it with return value optimization (RVO)?

In this case it looks like returning an invalid value is important, so maybe a simple `std::string` isn't expressive enough. I think this is a good use case for `std::optional<std::string>`, which is more expressive than a raw pointer:

```
    std::optional<std::string> Plugin::Name() const
    {
      if (!this->dataPtr->info)
        return std::nullopt;

      return {this->dataPtr->info->name};
    }
```

Comment (mxgrey) [reply core/include/ignition/plugin/Plugin.hh]: No, RVO doesn't really apply to the copying that we're talking about here.

If a `std::string` is returned by value, then the function requires a copy of the internally stored `name` field to be copied. This copy will be made whether or not the user has any need for a copy to be made. If they only want to view the name, then they have no use for making a copy, and it would be wasting time and resources (albeit, not very significant resources, so this discussion is reaching bike shedding status at this point).

A `const std::string*` is the perfect type expression for what I think this function should be providing, which is this: "Here is an immutable, non-owning view of the plugin's internally-stored name, or a `nullptr` if there is no plugin at all."

I understand that historically raw pointers have had some ambiguity to their meaning, but I think we're far enough into modern C++ that the ambiguity has died out (with a few straggling libraries like Qt). Besides, this meaning for raw pointers is used very consistently throughout the rest of `ignition-plugin`, so I don't see it being an issue here.

Comment (chapulina) [reply core/include/ignition/plugin/Plugin.hh]: Ok yeah you're right, RVO doesn't apply here.

I still think that we could leverage modern concepts that C++17 offers instead of potentially ambiguous raw pointers when introducing new API.

Comment (mxgrey) [reply core/include/ignition/plugin/Plugin.hh]: If we want to use the most correct modern C++ type expression, then we want a [`std::observer_ptr<std::string>`](https://en.cppreference.com/w/cpp/experimental/observer_ptr), but that's unfortunately not released in the Standard yet.

Until `std::observer_ptr` is available, raw pointers are the correct data type for providing a view of an object that might not be available (a reference would be preferable if we can guarantee the availability of the object, but obviously the tests revealed that we can't). Since `ignition-plugin` is a systems programming library, it would be bad practice to force a heap allocation and copy onto a user when they don't need it.

Comment (chapulina) [reply core/include/ignition/plugin/Plugin.hh]: > not released in the Standard yet

That's not an option then, right? :upside_down_face: 

> it would be bad practice to force a heap allocation and copy onto a user when they don't need it

Ok, fair enough, I can see how the pointer offers us a way to prevent a copy here. Hopefully we don't run into ownership issues and the caller remembers to make the copy if they need to hold that value beyond the lifetime of info.

Thank you for walking through this with me :bowing_woman:

File: core/include/ignition/plugin/Plugin.hh
Comment (chapulina) [inline core/include/ignition/plugin/Plugin.hh:141]: Mind documenting the `\return`?

**Review by chapulina — APPROVED** [review]
Thank you for iterating and for the comprehensive tests. I think this is good to go once the doxygen is updated.

File: core/include/ignition/plugin/Plugin.hh
Comment (nlamprian) [reply core/include/ignition/plugin/Plugin.hh]: Done. Updated doc of HasInterface above as well. It was the culprit here. Copy-pasting numbs your brain

**Comment by scpeters** [comment]
this is targeting `main`, so we can backport it to `ign-plugin1` after it is merged

**Comment by nlamprian** [comment]
> this is targeting `main`, so we can backport it to `ign-plugin1` after it is merged

Care to elaborate? I changed a header file, shouldn't this prevent it from being merged into ign-plugin1?

**Comment by scpeters** [comment]
> > this is targeting `main`, so we can backport it to `ign-plugin1` after it is merged
> 
> Care to elaborate? I changed a header file, shouldn't this prevent it from being merged into ign-plugin1?

adding a non-virtual function to a class doesn't break API or ABI, so it's fine to do for `ign-plugin1`. Per [semver.org](https://semver.org), we should bump the minor version when "adding functionality in a backwards compatible manner", but it doesn't break anything

**Comment by nlamprian** [comment]
In that case, I can simply rebase here. It doesn't seem there are any conflicts. Let me know.

**Comment by scpeters** [comment]
> In that case, I can simply rebase here. It doesn't seem there are any conflicts. Let me know.

sounds good. I just changed the target of this PR to `ign-plugin1`; please go ahead and rebase

**Review by scpeters — APPROVED** [review]
thanks for the contribution!

**Pull request merged**
